src/PPrintEngine.ml

```diff
@@ -58,39 +58,90 @@ class type output = object
 
 end
 
+(* ------------------------------------------------------------------------- *)
+
+(* Printing blank space (indentation characters). *)
+
+let blank_length =
+  80
+
+let blank_buffer =
+  String.make blank_length ' '
+
+let rec blanks output_fn n =
+  if n <= 0 then
+    ()
+  else if n <= blank_length then
+    output_fn blank_buffer 0 n
+  else begin
+    output_fn blank_buffer 0 blank_length;
+    blanks output_fn (n - blank_length)
+  end
+
+(* ------------------------------------------------------------------------- *)
+
+class with_blank_buffering char substring = object(self)
+  val mutable blank_buffered = None 
+
+  method private flush_blank =
+    match blank_buffered with
+    | None -> ()
+    | Some n -> 
+      blanks substring n;
+      blank_buffered <- None
+
+  method char c : unit =
+    begin match c with
+    | '\n' -> blank_buffered <- Some 0
+    | _ -> self#flush_blank
+    end;
+    char c
+
+  method substring s pos len =
+    if s == blank_buffer then (
+      blank_buffered <-
+        match blank_buffered with
+        | None -> Some len
+        | Some n -> Some (len + n)
+    ) else (
+      self#flush_blank;
+      substring s pos len
+    )
+end
+
 (* Three kinds of output channels are wrapped so as to satisfy the above
    interface: OCaml output channels, OCaml memory buffers, and OCaml
    formatters. *)
 
-class channel_output channel = object
-  method char = output_char channel
-  method substring = output_substring channel
+class channel_output channel =
```

Comment:
The class `channel_output` could be left unchanged (I hope), and combined with `with_blank_buffering` via some form of inheritance. (I am not sure of the details.)